src/ConstLinearLayer.jl

```diff
@@ -2,17 +2,11 @@ import ChainRulesCore: rrule
 using LuxCore
 using LuxCore: AbstractExplicitLayer
 
-struct ConstLinearLayer{T} <: AbstractExplicitLayer # where {in_dim,out_dim,T}
-    W::AbstractMatrix{T}  
-    position::Union{Vector{Int64}, UnitRange{Int64}}
-    in_dim::Integer
-    out_dim::Integer
+struct ConstLinearLayer <: AbstractExplicitLayer
+    op
 end
 
-ConstLinearLayer(W::AbstractMatrix{T}) where T = ConstLinearLayer(W,1:size(W,2),size(W,2),size(W,1))
-ConstLinearLayer(W::AbstractMatrix{T}, pos::Union{Vector{Int64}, UnitRange{Int64}}) where T = ConstLinearLayer(W,pos,size(W,2),size(W,1))
-
-(l::ConstLinearLayer)(x::AbstractVector) = l.in_dim == length(x[l.position]) ? l.W * x[l.position] : error("x (or the position index) has a wrong length!")
+(l::ConstLinearLayer)(x::AbstractVector) = l.op * x[1:size(l.op,2)]
```

Comment:
why are you using `1:size(l.op, 2)` here? If `x` has the wrong size then this operation should fail I think. (What's the application of this?)

Reply:
I was thinking that if some AA basis is not used, it might be ignorable. But yes, I agree that as a general ConstLinearLayer, its operation should not be defined like this. Will modify it soon.

Reply:
You can always compose several linear maps using LinearOperators

Reply:
Thanks for pointing that out. But currently I just extend the $L$-th A2Bmap (a # $BB_L\times$ # $AA_L$ sparse matrix) to a # $BB_L\times$ # $allAA$ sparse matrix - not much linear operator is needed. So perhaps we still don't need to rely on LinearOperators.jl for now?

Reply:
And in this way, the above operation is just l(x) = l.op * x now

Reply:
> You can always compose several linear maps using LinearOperators

Oh, perhaps I get what you mean. I will try another implementation with LinearOperators and do a quick comparison.

Reply:
I checked the following:

let $op$ be one of the A2Bmap, $SL$ the sparse linear transformation that contracts x to x[pos] , and  run

```julia
@btime $op * $SL * $x;
@btime LinearOperator($op) * LinearOperator($SL) * $x;
@btime $op * $x[$pos];
```

It turns out:
```julia
9.917 μs (2 allocations: 7.53 KiB)
11.833 μs (37 allocations: 21.59 KiB)
1.492 μs (2 allocations: 7.53 KiB)
```

I could have misunderstood something, but maybe the linear operators involved here are so sparse that avoiding that matrix-matrix multiplication was not so necessary, Also, perhaps the `pos` idea (the third one) wasn't too bad?

Reply:
Just recording a more thorough test:

```julia
using RepLieGroups, EquivariantModels, Polynomials4ML
using EquivariantModels: degord2spec, simple_radial_basis, xx2AA, _rpi_A2B_matrix, RPE_filter, sparse_trans

# Constructing an AAspec
totdeg = 6
ν = 2
Lmax = 2
basis = legendre_basis(totdeg)
radial = EquivariantModels.simple_radial_basis(basis)
Aspec, AAspec = degord2spec(radial; totaldegree = totdeg, 
                              order = ν, 
                              Lmax = Lmax, islong = true)

# Generate a random configure X and compute `AA = AA(X)`
l_basis, ps_basis, st_basis = xx2AA(AAspec, radial)
X = [ @SVector(rand(3)) for i in 1:10 ]
AA = l_basis(X, ps_basis, st_basis)[1]

# Now when we want an L = 1 block, only certain AA elements are needed
L = 1
cgen = RepLieGroups.O3.Rot3DCoeffs(L)
pos = findall(x -> x==1, RPE_filter(L).(AAspec))
ST = sparse_trans(pos,length(AAspec)) # A sparse transformation(ST) that projects AA to AA[pos]
C = _rpi_A2B_matrix(cgen, AAspec[pos]) # Coupling coefficient matrix

using BenchmarkTools

C * ST * AA == C * (ST * AA) == (C * ST) * A == C * AA[pos]

@btime C * ST * A;
@btime C * (ST * A);
@btime (C * ST) * A;
@btime C * A[pos];

TRAN = C * ST
TRAN * AA == C * AA[pos]
@btime C * ST;
@btime TRAN * AA;
@btime C * AA[pos];
```

its output reads
```julia
  336.333 μs (2 allocations: 6.80 KiB)
  336.375 μs (2 allocations: 6.80 KiB)
  85.416 μs (6 allocations: 24.56 KiB)
  1.092 μs (2 allocations: 6.80 KiB)
  4.993 μs (5 allocations: 22.58 KiB)
  80.083 μs (1 allocation: 1.98 KiB)
  1.104 μs (2 allocations: 6.80 KiB)
```
which still indicates that the "keep the position" strategy is not too bad. However, this is at the cost of doing a filtering for each L - I will do one more round of comparison.

Reply:
I tried the following:
```julia
## Test whether or not a filtering is needed
# given an AA basis, we compare the following three ways of computing L=0,1,...,Lmax BB basis

# way 1: throw the whole AAspec to each L
function f1(AA,Lmax)
    for L = 0:Lmax
        cgen = RepLieGroups.O3.Rot3DCoeffs(L)
        C = _rpi_A2B_matrix(cgen, AAspec)
        C * AA
    end
end

# way 2: throw the filtered AAspec to each L
function f2(AA,Lmax)
    for L = 0:Lmax
        cgen = RepLieGroups.O3.Rot3DCoeffs(L)
        pos = findall(x -> x==1, RPE_filter(L).(AAspec))
        C = _rpi_A2B_matrix(cgen, AAspec[pos])
        C * AA[pos]
    end
end

# way 3: throw the filtered AAspec to each L and use sparse transformation
function f3(AA,Lmax)
    for L = 0:Lmax
        cgen = RepLieGroups.O3.Rot3DCoeffs(L)
        pos = findall(x -> x==1, RPE_filter(L).(AAspec))
        ST = sparse_trans(pos,length(AAspec)) # generates a linear projector from AA to AA[pos]
        C = _rpi_A2B_matrix(cgen, AAspec[pos])
        (C * ST) * AA
    end
end
f1(AA,Lmax)
f2(AA,Lmax)
f3(AA,Lmax)
@btime f1(AA,Lmax)
@btime f2(AA,Lmax)
@btime f3(AA,Lmax)
```

which turns out
```julia
  23.615 s (424138663 allocations: 12.82 GiB)
  21.460 s (390889001 allocations: 10.39 GiB)
  21.475 s (390889127 allocations: 10.55 GiB)
```

So it suggests that filtering for each L might be a "must" (kind of intuitive), and the "position" strategy is still the fastest, though the difference between the latter two is almost negligible. I am hence happy with both of the latter (will keep the f3 way for now).

Reply:
Thanks for looking into this. I'm finding it very hard to believe. I will rerun this to make sure.

Reply:
note also you are still constructing the matrix on the fly...

Reply:
So I rand some small tests. First of all you partially right - the performance of the large sparse matrix is abysmal. But one can still fix this with a `LinearOperator`. Here is a prototype of what I have in mind: 
```julia
julia> using SparseArrays, LinearAlgebra, BenchmarkTools, LinearOperators
       C = sprandn(100, 300, 0.01)
       AA = randn(3000)
       idx = 1:300
       P = sparse(I, 300, 3000)
       CP = C * P
       fL = let C=C, idx=idx
               (res, aa) -> mul!(res, C, aa[idx])
            end
       L = LinearOperator(Float64, 100, 3000, false, false, fL)
       @btime $C * ($AA)[($idx)]
       @btime $CP * $AA
       @btime $L * $AA;
  471.092 ns (2 allocations: 3.38 KiB)
  1.986 μs (1 allocation: 896 bytes)
  475.128 ns (2 allocations: 3.38 KiB)
```

Reply:
Thanks so much for looking into this! I am on it